subsys/debug/tracing/cpu_stats.c

```diff
@@ -0,0 +1,209 @@
+/*
+ * Copyright (c) 2018 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <tracing_cpu_stats.h>
+
+enum cpu_state {
+	CPU_STATE_IDLE,
+	CPU_STATE_NON_IDLE,
+	CPU_STATE_SCHEDULER
+};
+
+static enum cpu_state last_cpu_state = CPU_STATE_SCHEDULER;
+static enum cpu_state cpu_state_before_interrupts;
+
+static u32_t last_time;
+static u64_t idle_cnt;
+static u64_t non_idle_cnt;
+static u64_t sched_cnt;
+static int nested_interrupts;
+static struct k_thread *current_thread;
+
+#ifndef CONFIG_SMP
+extern k_tid_t const _idle_thread;
+#endif
+
+static inline int is_idle_thread(struct k_thread *thread)
+{
+#ifdef CONFIG_SMP
+	return thread->base.is_idle;
+#else
+	return thread == _idle_thread;
+#endif
+}
+
+void increase_counter(volatile u64_t *cnt)
+{
+	int key = irq_lock();
+	u32_t time = k_cycle_get_32();
+
+	if (time >= last_time) {
+		(*cnt) += (time - last_time);
+	} else {
+		(*cnt) += (UINT32_MAX - last_time + 1 + time);
+	}
+	last_time = time;
+	irq_unlock(key);
+}
+
+static void cpu_stats_update_counters(void)
+{
+	switch (last_cpu_state) {
+	case CPU_STATE_IDLE:
+		increase_counter(&idle_cnt);
+		break;
+
+	case CPU_STATE_NON_IDLE:
+		increase_counter(&non_idle_cnt);
+		break;
+
+	case CPU_STATE_SCHEDULER:
+		increase_counter(&sched_cnt);
+		break;
+
+	default:
+		__ASSERT_NO_MSG(false);
+		break;
+	}
+}
+
+u64_t cpu_stats_sched_get_ns(void)
+{
+	cpu_stats_update_counters();
+	return SYS_CLOCK_HW_CYCLES_TO_NS(sched_cnt);
+}
+
+u64_t cpu_stats_idle_get_ns(void)
+{
+	cpu_stats_update_counters();
+	return SYS_CLOCK_HW_CYCLES_TO_NS(idle_cnt);
+}
+
+u64_t cpu_stats_non_idle_get_ns(void)
+{
+	cpu_stats_update_counters();
+	return SYS_CLOCK_HW_CYCLES_TO_NS(non_idle_cnt);
+}
+
+u32_t cpu_stats_non_idle_percent_get(void)
+{
+	cpu_stats_update_counters();
+	return ((non_idle_cnt + sched_cnt) * 100) /
+		(idle_cnt + non_idle_cnt + sched_cnt);
+}
+
+void cpu_stats_reset_counters(void)
+{
+	int key = irq_lock();
+
+	idle_cnt = 0;
+	non_idle_cnt = 0;
+	sched_cnt = 0;
+	last_time = k_cycle_get_32();
+	irq_unlock(key);
+}
+
+void sys_trace_thread_switched_in(void)
+{
+	__ASSERT_NO_MSG(nested_interrupts == 0);
```

Comment:
I would lock it too. Otherwise interrupt can come and harm us. (Even if upper layer disables interrupts, if we depend on them being disabled we should either disable them or document the requirement and assert state). I prefer adding a lock - it won't increase the cost of call much as lock is already taken when counters are updated.

Reply:
done